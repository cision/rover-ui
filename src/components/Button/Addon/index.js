import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import style from './style.css';

const Addon = ({ className, size, ...passedProps }) => (
  <div
    {...passedProps}
    className={classNames(className, style.Addon, style[size])}
  />
);

Addon.propTypes = {
  className: PropTypes.string,
  size: PropTypes.oneOf(['sm', 'md', 'lg']).isRequired,
};

Addon.defaultProps = {
  className: '',
};

export default Addon;
