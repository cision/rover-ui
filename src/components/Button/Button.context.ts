import { createContext } from 'react';

type ContextProps = {
  size?: string;
};

export default createContext<ContextProps>({});
