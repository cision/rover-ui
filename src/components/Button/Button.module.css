.Button {
  -webkit-appearance: none;
  background: transparent none;
  border: var(--rvr-border-width-md) solid transparent;
  box-sizing: border-box;
  cursor: pointer;
  font-size: var(--rvr-font-size-md);
  line-height: var(--rvr-line-height-sm);
  transition: 0.2s linear background-color, 0.2s linear border-color, 0.2s linear color;
  display: inline-flex;
  flex-flow: row nowrap;
  align-content: center;
  align-items: center;
  text-decoration: none;
}

.Button.disabled,
.Button:disabled,
.Button[disabled] {
  cursor: default;
}

.buttonBody {
  flex: 1 1 auto;
  min-width: 0;
}

/* Sizes */

.sm {
  padding: var(--rvr-space-bordered-xs) var(--rvr-space-bordered-sm);
  border-radius: calc(var(--rvr-space-bordered-sm) + (var(--rvr-line-height-md) / 2));
}

.md {
  padding: var(--rvr-space-bordered-sm) var(--rvr-space-bordered-md);
  border-radius: calc(var(--rvr-space-bordered-md) + (var(--rvr-line-height-md) / 2));
}

.lg {
  padding: var(--rvr-space-bordered-md) var(--rvr-space-bordered-lg);
  border-radius: calc(var(--rvr-space-bordered-lg) + (var(--rvr-line-height-md) / 2));
}

/* Circle */
.circle {
  display: inline-block;
  text-align: center;
}

.sm.circle {
  height: calc(var(--rvr-space-bordered-xs) * 2 + var(--rvr-line-height-md));
  width: calc(var(--rvr-space-bordered-xs) * 2 + var(--rvr-line-height-md));
  padding: var(--rvr-space-bordered-xs);
  border-radius: calc(var(--rvr-space-bordered-sm) + (var(--rvr-line-height-md) / 2));
}

.md.circle {
  height: calc(var(--rvr-space-bordered-sm) * 2 + var(--rvr-line-height-md));
  width: calc(var(--rvr-space-bordered-sm) * 2 + var(--rvr-line-height-md));
  padding: var(--rvr-space-bordered-sm);
  border-radius: calc(var(--rvr-space-bordered-md) + (var(--rvr-line-height-md) / 2));
}

.lg.circle {
  height: calc(var(--rvr-space-bordered-md) * 2 + var(--rvr-line-height-md));
  width: calc(var(--rvr-space-bordered-md) * 2 + var(--rvr-line-height-md));
  padding: var(--rvr-space-bordered-md);
  border-radius: calc(var(--rvr-space-bordered-md) + (var(--rvr-line-height-md) / 2));
}

/* Button level colors */

.primary,
.primary:link,
.primary:visited {
  color: var(--rvr-btn-color-primary);
  background-color: var(--rvr-btn-bg-color-primary);
  border-color: var(--rvr-btn-border-color-primary);
}

.primary.active,
.primary:active,
.primary.hover,
.primary:hover,
.primary.focus,
.primary:focus {
  color: var(--rvr-btn-color-primary-active);
  background-color: var(--rvr-btn-bg-color-primary-active);
  border-color: var(--rvr-btn-border-color-primary-active);
}

.primary.disabled,
.primary:disabled,
.primary[disabled] {
  color: var(--rvr-btn-color-primary-disabled);
  background-color: var(--rvr-btn-bg-color-primary-disabled);
  border-color: var(--rvr-btn-border-color-primary-disabled);
}

.primary-alt,
.primary-alt:link,
.primary-alt:visited {
  color: var(--rvr-btn-color-primary-alt);
  background-color: var(--rvr-btn-bg-color-primary-alt);
  border-color: var(--rvr-btn-border-color-primary-alt);
}

.primary-alt.active,
.primary-alt:active,
.primary-alt.hover,
.primary-alt:hover,
.primary-alt.focus,
.primary-alt:focus {
  color: var(--rvr-btn-color-primary-alt-active);
  background-color: var(--rvr-btn-bg-color-primary-alt-active);
  border-color: var(--rvr-btn-border-color-primary-alt-active);
}

.primary-alt.disabled,
.primary-alt:disabled,
.primary-alt[disabled] {
  color: var(--rvr-btn-color-primary-alt-disabled);
  background-color: var(--rvr-btn-bg-color-primary-alt-disabled);
  border-color: var(--rvr-btn-border-color-primary-alt-disabled);
}

.secondary,
.secondary:link,
.secondary:visited {
  color: var(--rvr-btn-color-secondary);
  background-color: var(--rvr-btn-bg-color-secondary);
  border-color: var(--rvr-btn-border-color-secondary);
}

.secondary.active,
.secondary:active,
.secondary.hover,
.secondary:hover,
.secondary.focus,
.secondary:focus {
  color: var(--rvr-btn-color-secondary-active);
  background-color: var(--rvr-btn-bg-color-secondary-active);
  border-color: var(--rvr-btn-border-color-secondary-active);
}

.secondary.disabled,
.secondary:disabled,
.secondary[disabled] {
  color: var(--rvr-btn-color-secondary-disabled);
  background-color: var(--rvr-btn-bg-color-secondary-disabled);
  border-color: var(--rvr-btn-border-color-secondary-disabled);
}

.tertiary,
.tertiary:link,
.tertiary:visited {
  color: var(--rvr-btn-color-tertiary);
  background-color: var(--rvr-btn-bg-color-tertiary);
  border-color: var(--rvr-btn-border-color-tertiary);
}

.tertiary.active,
.tertiary:active,
.tertiary.hover,
.tertiary:hover,
.tertiary.focus,
.tertiary:focus {
  color: var(--rvr-btn-color-tertiary-active);
  background-color: var(--rvr-btn-bg-color-tertiary-active);
  border-color: var(--rvr-btn-border-color-tertiary-active);
}

.tertiary.disabled,
.tertiary:disabled,
.tertiary[disabled] {
  color: var(--rvr-btn-color-tertiary-disabled);
  background-color: var(--rvr-btn-bg-color-tertiary-disabled);
  border-color: var(--rvr-btn-border-color-tertiary-disabled);
}

.success,
.success:link,
.success:visited {
  color: var(--rvr-btn-color-success);
  background-color: var(--rvr-btn-bg-color-success);
  border-color: var(--rvr-btn-border-color-success);
}

.success.active,
.success:active,
.success.hover,
.success:hover,
.success.focus,
.success:focus {
  color: var(--rvr-btn-color-success-active);
  background-color: var(--rvr-btn-bg-color-success-active);
  border-color: var(--rvr-btn-border-color-success-active);
}

.success.disabled,
.success:disabled,
.success[disabled] {
  color: var(--rvr-btn-color-primary-alt-disabled);
  background-color: var(--rvr-btn-bg-color-primary-alt-disabled);
  border-color: var(--rvr-btn-border-color-primary-alt-disabled);
}

.danger,
.danger:link,
.danger:visited {
  color: var(--rvr-btn-color-danger);
  background-color: var(--rvr-btn-bg-color-danger);
  border-color: var(--rvr-btn-border-color-danger);
}

.danger.active,
.danger:active,
.danger.hover,
.danger:hover,
.danger.focus,
.danger:focus {
  color: var(--rvr-btn-color-danger-active);
  background-color: var(--rvr-btn-bg-color-danger-active);
  border-color: var(--rvr-btn-border-color-danger-active);
}

.danger.disabled,
.danger:disabled,
.danger[disabled] {
  color: var(--rvr-btn-color-primary-alt-disabled);
  background-color: var(--rvr-btn-bg-color-primary-alt-disabled);
  border-color: var(--rvr-btn-border-color-primary-alt-disabled);
}

.link,
.link:link,
.link:visited {
  color: var(--rvr-btn-color-link);
  background-color: var(--rvr-btn-bg-color-link);
  border-color: var(--rvr-btn-border-color-link);
}

.link.active,
.link:active,
.link.hover,
.link:hover,
.link.focus,
.link:focus {
  color: var(--rvr-btn-color-link-active);
  background-color: var(--rvr-btn-bg-color-link-active);
  border-color: var(--rvr-btn-border-color-link-active);
}

.link.disabled,
.link:disabled,
.link[disabled] {
  color: var(--rvr-btn-color-link-disabled);
  background-color: var(--rvr-btn-bg-color-link-disabled);
  border-color: var(--rvr-btn-border-color-link-disabled);
}

.text,
.text:link,
.text:visited {
  color: var(--rvr-btn-color-text);
  background-color: var(--rvr-btn-bg-color-text);
  border-color: var(--rvr-btn-border-color-text);
}

.text.active,
.text:active,
.text.hover,
.text:hover,
.text.focus,
.text:focus {
  color: var(--rvr-btn-color-text-active);
  background-color: var(--rvr-btn-bg-color-text-active);
  border-color: var(--rvr-btn-border-color-text-active);
}

.text.disabled,
.text:disabled,
.text[disabled] {
  color: var(--rvr-btn-color-text-disabled);
  background-color: var(--rvr-btn-bg-color-text-disabled);
  border-color: var(--rvr-btn-border-color-text-disabled);
}

.hollow,
.hollow:link,
.hollow:visited {
  color: var(--rvr-btn-color-hollow);
  background-color: var(--rvr-btn-bg-color-hollow);
  border-color: var(--rvr-btn-border-color-hollow);
}

.hollow.active,
.hollow:active,
.hollow.hover,
.hollow:hover,
.hollow.focus,
.hollow:focus {
  color: var(--rvr-btn-color-hollow-active);
  background-color: var(--rvr-btn-bg-color-hollow-active);
  border-color: var(--rvr-btn-border-color-hollow-active);
}

.hollow.disabled,
.hollow:disabled,
.hollow[disabled] {
  color: var(--rvr-btn-color-hollow-disabled);
  background-color: var(--rvr-btn-bg-color-hollow-disabled);
  border-color: var(--rvr-btn-border-color-hollow-disabled);
}
