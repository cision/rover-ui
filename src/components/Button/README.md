# Button

### Button component is used for all links and calls to action, as well as triggers for dropdowns

The `<Button/>` component can wrap any kind of node.

By default, it creates a `<button>` tag, and it propagates down all
the props it doesn't use for appearance.

That means it will propagate all DOM events that react supports, like
`onClick`, `onFocus`, etc.

You can change the tag name (aka DOM element) to something like `a`
or `div`. If you do, React will allow different props. It's up to
you to not put an `href` on a `button`.
