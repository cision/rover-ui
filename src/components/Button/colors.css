:root {
  /* PRIMARY */
  --color-primary: var(--white);
  --bgColor-primary: var(--blue);
  --borderColor-primary: var(--bgColor-primary);

  /* Active */
  --color-primary-active: var(--color-primary);
  --bgColor-primary-active: var(--blue-lite-1);
  --borderColor-primary-active: var(--bgColor-primary-active);

  /* Disabled */
  --color-primary-disabled: var(--white);
  --bgColor-primary-disabled: var(--gray-lite-3);
  --borderColor-primary-disabled: var(--gray-lite-3);

  /* TEAL */
  --color-teal: var(--white);
  --bgColor-teal: var(--teal);
  --borderColor-teal: var(--bgColor-teal);

  /* Active */
  --color-teal-active: var(--color-teal);
  --bgColor-teal-active: var(--teal-lite-1);
  --borderColor-teal-active: var(--bgColor-teal-active);

  /* Disabled */
  --color-teal-disabled: var(--color-primary-disabled);
  --bgColor-teal-disabled: var(--bgColor-primary-disabled);
  --borderColor-teal-disabled: var(--borderColor-primary-disabled);

  /* SECONDARY */
  --color-secondary: var(--blue);
  --bgColor-secondary: var(--white);
  --borderColor-secondary: var(--color-secondary);

  /* Active */
  --color-secondary-active: var(--color-primary-active);
  --bgColor-secondary-active: var(--bgColor-primary-active);
  --borderColor-secondary-active: var(--borderColor-primary-active);

  /* Disabled */
  --color-secondary-disabled: var(--gray-lite-3);
  --bgColor-secondary-disabled: var(--gray-lite-6);
  --borderColor-secondary-disabled: var(--gray-lite-4);

  /* SUCCESS */
  --color-success: var(--white);
  --bgColor-success: var(--green);
  --borderColor-success: var(--bgColor-success);

  /* Active */
  --color-success-active: var(--white);
  --bgColor-success-active: var(--green-lite-1);
  --borderColor-success-active: var(--bgColor-success-active);

  /* Disabled */
  --color-success-disabled: var(--color-secondary-disabled);
  --bgColor-success-disabled: var(--bgColor-secondary-disabled);
  --borderColor-success-disabled: var(--borderColor-secondary-disabled);

  /* TERTIARY */
  --color-tertiary: var(--gray-lite-2);
  --bgColor-tertiary: var(--white);
  --borderColor-tertiary: var(--gray-lite-3);

  /* Active */
  --color-tertiary-active: var(--gray);
  --bgColor-tertiary-active: var(--gray-lite-6);
  --borderColor-tertiary-active: var(--gray-lite-2);

  /* Disabled */
  --color-tertiary-disabled: var(--color-secondary-disabled);
  --bgColor-tertiary-disabled: var(--bgColor-secondary-disabled);
  --borderColor-tertiary-disabled: var(--borderColor-secondary-disabled);

  /* DARK MODE */
  --color-darkMode: var(--white);
  --bgColor-darkMode: transparent;
  --borderColor-darkMode: var(--white);

  /* Active */
  --color-darkMode-active: var(--gray-lite-3);
  --bgColor-darkMode-active: var(--bgColor-darkMode);
  --borderColor-darkMode-active: var(--gray-lite-3);

  /* Disabled */
  --color-darkMode-disabled: #fff6;
  --bgColor-darkMode-disabled: var(--bgColor-darkMode);
  --borderColor-darkMode-disabled: #fff6;
}
