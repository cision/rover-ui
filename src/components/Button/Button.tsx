import React, { useMemo } from 'react';
import classNames from 'classnames';

import Addon, { AddonProps } from './Addon/Addon';
import styles from './Button.module.css';
import { TButtonLevel, TButtonSize, TButtonState } from './types';

/*
  These 2 variants are in design docs but might not be "Buttons" in a strict sense.
  Filter (-> Should probably be a pill)
  Select (-> These might be better served as a different component)
*/

interface BaseButtonProps {
  circle?: boolean;
  className?: string;
  darkMode?: boolean;
  hollow?: boolean;
  level?: TButtonLevel;
  size?: TButtonSize;
  tag?: React.ComponentType<{ [key: string]: unknown }> | null;

  // States
  hover?: boolean;
  active?: boolean;
  disabled?: boolean;
  checked?: boolean;
  focus?: boolean;
}

// Button props
type ButtonElementProps = React.ButtonHTMLAttributes<HTMLButtonElement> &
  BaseButtonProps & {
    href?: undefined;
  };

// Anchor props
type AnchorElementProps = React.AnchorHTMLAttributes<HTMLAnchorElement> &
  BaseButtonProps & {
    href?: string;
  };

type ButtonType = React.FC<ButtonElementProps | AnchorElementProps> & {
  Addon: React.FC<AddonProps>;
};

// Guard to check if href exists in props
const hasHref = (
  props: ButtonElementProps | AnchorElementProps
): props is AnchorElementProps => 'href' in props && props.href !== undefined;

// Design doc names in comments
export const levels: TButtonLevel[] = [
  'primary', // "Primary"
  'primary-alt', // "Teal"
  'secondary', // "Secondary"
  'tertiary', // "Tertiary"
  'success', // "Green"
  'danger', // "Red"
  'link', // "Blue"
  'text', // "Gray"
];

// These states are each their own boolean prop
export const states: TButtonState[] = [
  'hover', // "Hover"
  'focus', // "Focus"
  'active', // "Pressed"
  'disabled', // "Disabled"
  'checked', // "Active"
];

export const sizes: TButtonSize[] = [
  'sm', // ""
  'md', // "Medium"
  'lg', // "Large"
];

const Button: ButtonType = ({
  children: initChildren = null,
  circle = undefined,
  className: passedClassName = '',
  darkMode = false,
  hollow = false,
  level = 'secondary',
  size = 'lg',
  tag: Tag = null,
  ...passedProps
}: ButtonElementProps | AnchorElementProps) => {
  const children = initChildren;

  // For future ref
  const truthyStateKeys = useMemo(
    () => states.filter((state) => !!passedProps[state]),
    [passedProps]
  );

  // Filter out undefined passedProps and `active` from DOM element tags
  // const safePassedProps = useMemo(
  //   () =>
  //     Object.entries(passedProps).reduce((result, [key, value]) => {
  //       if (value === undefined) {
  //         return result;
  //       }

  //       if (typeof Tag === 'string' && key === 'active') {
  //         return result;
  //       }

  //       return { ...result, [key]: value };
  //     }, {}),
  //   [passedProps, Tag]
  // );

  const className = useMemo(
    () =>
      classNames(
        passedClassName,
        styles.Button,
        styles[level],
        styles[size],
        {
          [styles.hollow]: hollow || darkMode,
          [styles.circle]: circle,
          [styles['primary-alt']]: level === 'teal', // For backwards compatibility with the old level name
        },
        /*
          In addition to "real" boolean props, this adds class names for
          'disabled', 'active', etc. so consumers can force appearance on
          elements easily
        */
        truthyStateKeys.map((truthyStateKey) => styles[truthyStateKey])
      ),
    [circle, darkMode, hollow, level, passedClassName, size, truthyStateKeys]
  );

  if (Tag) {
    return (
      <Tag {...passedProps} className={className}>
        {children}
      </Tag>
    );
  }

  if (hasHref(passedProps)) {
    return (
      <a {...passedProps} className={className}>
        {children}
      </a>
    );
  }

  // button render
  return (
    <button
      {...(passedProps as ButtonElementProps)}
      type="button"
      className={className}
    >
      {children}
    </button>
  );
};

Button.Addon = Addon;

export default Button;
