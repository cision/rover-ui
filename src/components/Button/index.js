import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { propTypes as tagPropTypes } from '../../shared/models/tag';

import Addon from './Addon';
import styles from './Button.module.css';

/*
  These 2 variants are in design docs but might not be "Buttons" in a strict sense.
  Filter (-> Should probably be a pill)
  Select (-> These might be better served as a different component)
*/

// Design doc names in comments
export const levels = [
  'primary', // "Primary"
  'primary-alt', // "Teal"
  'secondary', // "Secondary"
  'tertiary', // "Tertiary"
  'success', // "Green"
  'danger', // "Red"
  'link', // "Blue"
  'text', // "Gray"
];

// These states are each their own boolean prop
export const states = [
  'hover', // "Hover"
  'focus', // "Focus"
  'active', // "Pressed"
  'disabled', // "Disabled"
  'checked', // "Active"
];

export const sizes = [
  'sm', // ""
  'md', // "Medium"
  'lg', // "Large"
];

const Button = (props) => {
  const {
    children: initChildren,
    circle,
    className: passedClassName,
    darkMode,
    hollow,
    level,
    size,
    tag: Tag,
    ...passedProps
  } = props;

  let children = initChildren;

  // For future ref
  const truthyStateKeys = useMemo(
    () => states.filter((state) => !!passedProps[state]),
    [passedProps]
  );

  // Filter out undefined passedProps and `active` from DOM element tags
  const safePassedProps = useMemo(
    () =>
      Object.entries(passedProps).reduce((result, [key, value]) => {
        if (value === undefined) {
          return result;
        }

        if (typeof Tag === 'string' && key === 'active') {
          return result;
        }

        return { ...result, [key]: value };
      }, {}),
    [passedProps, Tag]
  );

  const className = useMemo(
    () =>
      classNames(
        passedClassName,
        styles.Button,
        styles[level],
        styles[size],
        {
          [styles.hollow]: hollow || darkMode,
          [styles.circle]: circle,
          [styles['primary-alt']]: level === 'teal', // For backwards compatibility with the old level name
        },
        /*
          In addition to "real" boolean props, this adds class names for
          'disabled', 'active', etc. so consumers can force appearance on
          elements easily
        */
        truthyStateKeys.map((truthyStateKey) => styles[truthyStateKey])
      ),
    [circle, darkMode, hollow, level, passedClassName, size, truthyStateKeys]
  );

  const addonChildren = React.Children.toArray(initChildren).filter(
    (child) =>
      child && child.type && child.type.displayName === Addon.displayName
  );

  if (addonChildren.length) {
    children = React.Children.map(initChildren, (child) => {
      if (child && child.type && child.type.displayName === Addon.displayName) {
        return React.cloneElement(child, { ...child.props, size });
      }

      if (typeof child === 'string') {
        return <span className={styles.buttonBody}>{child}</span>;
      }

      return child;
    });
  }

  return (
    <Tag {...safePassedProps} className={className}>
      {children}
    </Tag>
  );
};

Button.propTypes = {
  active: PropTypes.bool,
  children: PropTypes.node,
  /** `circle` buttons have a fixed width === height */
  circle: PropTypes.bool,
  className: PropTypes.string,
  /** `darkMode` is deprecated in favor of `hollow` */
  darkMode: PropTypes.bool,
  /** `hollow` buttons have an outline style, and are suitable for dark backgrounds */
  hollow: PropTypes.bool,
  level: PropTypes.oneOf(levels),
  size: PropTypes.oneOf(sizes),
  tag: tagPropTypes,
  ...states.reduce(
    (result, stateKey) => ({
      ...result,
      [stateKey]: PropTypes.bool,
    }),
    {}
  ),
};

Button.defaultProps = {
  active: false,
  children: null,
  circle: undefined,
  className: '',
  darkMode: false,
  hollow: false,
  level: 'secondary',
  size: 'lg',
  tag: 'button',
  ...states.reduce(
    (result, stateKey) => ({
      ...result,
      [stateKey]: undefined,
    }),
    {}
  ),
};

Button.Addon = Addon;

export default Button;
