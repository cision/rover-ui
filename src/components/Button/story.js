import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean, text, select } from '@storybook/addon-knobs';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Button from './';
import ButtonReadme from './README.md';
import AddonReadme from './Addon/README.md';

const levels = ['tertiary', 'teal', 'secondary', 'primary'];
const sizes = ['sm', 'md', 'lg'];

const MyComponent = props => (
  <button {...props} className={classNames(props.className, 'MyComponent')} />
);

MyComponent.propTypes = { className: PropTypes.string };
MyComponent.defaultProps = { className: '' };

const tags = ['a', 'button', '<MyComponent />'];

storiesOf('Planets/Button', module)
  .addParameters({
    readme: {
      sidebar: ButtonReadme,
    },
  })
  .add(
    'Overview',
    () => (
      <Button
        className={text('className', '')}
        darkMode={boolean('darkMode', false)}
        disabled={boolean('disabled', false)}
        href={select('tag', tags, 'button') === 'a' ? '#' : null}
        level={select('level', levels, 'secondary')}
        onClick={action('Button clicked')}
        size={select('size', sizes, 'lg')}
        tag={
          select('tag', tags, 'button') === '&lt;MyComponent /&gt;'
            ? MyComponent
            : select('tag', tags, 'button')
        }
      >
        {text('children', 'Click me!')}
      </Button>
    ),
    {
      info: {
        inline: false,
        source: true,
      },
    }
  );

storiesOf('Planets/Button/Addon', module)
  .addParameters({
    readme: {
      sidebar: AddonReadme,
    },
  })
  .add(
    'Overview',
    () => (
      <Button size={select('<Button/> size', sizes, 'lg')}>
        <Button.Addon className={text('className', '')}>
          {text('children', 'ðŸ˜¸')}
        </Button.Addon>
        Click me!
      </Button>
    ),
    {
      info: {
        inline: false,
        source: true,
      },
    }
  );
