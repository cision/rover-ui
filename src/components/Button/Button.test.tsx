import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Button from '.';

describe('Button', () => {
  it('renders a `button` tag without error', () => {
    render(<Button />);
    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('You can override default props', () => {
    render(<Button type="submit">Boom</Button>);
    expect(screen.getByRole('button').getAttribute('type')).toEqual('submit');
  });

  describe('props.children', () => {
    it('renders strings', () => {
      render(<Button className="foo">Boom</Button>);
      expect(screen.getByRole('button').textContent).toEqual('Boom');
    });

    it('renders React nodes', () => {
      render(
        <Button className="foo">
          <span data-testid="bar">Boom</span>
        </Button>
      );

      expect(screen.getByRole('button').textContent).toEqual('Boom');
      expect(screen.getByTestId('bar')).toBeTruthy();
    });
  });

  describe('props.className', () => {
    it('adds custom className', () => {
      render(<Button className="foo">Boom</Button>);
      expect(screen.getByRole('button').classList).toContain('foo');
    });
  });

  describe('props.level', () => {
    it('adds custom className', () => {
      render(<Button level="primary">Boom</Button>);
      expect(screen.getByRole('button').classList).toContain('primary');
    });
  });

  describe('props.size', () => {
    describe('is available to `Addon` children by default', () => {
      it('sm', () => {
        render(
          <Button size="sm">
            <Button.Addon>Addon</Button.Addon>
          </Button>
        );

        expect(screen.getByText('Addon').classList).toContain('sm');
      });

      it('lg', () => {
        render(
          <Button size="lg">
            <Button.Addon>Addon</Button.Addon>
          </Button>
        );

        expect(screen.getByText('Addon').classList).toContain('lg');
      });
    });

    describe('can be overridden by `Addon` children', () => {
      it('sm', () => {
        render(
          <Button size="sm">
            <Button.Addon size="md">Addon</Button.Addon>
          </Button>
        );

        expect(screen.getByText('Addon').classList).toContain('md');
      });

      it('lg', () => {
        render(
          <Button size="lg">
            <Button.Addon size="md">Addon</Button.Addon>
          </Button>
        );

        expect(screen.getByText('Addon').classList).toContain('md');
      });
    });
  });

  describe('props.href', () => {
    it('renders an anchor tag when there is an href', () => {
      render(<Button href="#example">Boom</Button>);
      expect(screen.getByRole('link')).toBeTruthy();
    });

    it('renders a button tag when there is not href', () => {
      const callback = jest.fn();

      render(<Button onClick={callback}>Boom</Button>);

      const button = screen.getByRole('button');
      userEvent.click(button);

      expect(button).toBeTruthy();
      expect(callback).toBeCalledTimes(1);
    });
  });
});
