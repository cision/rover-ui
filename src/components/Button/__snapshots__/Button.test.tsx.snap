// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a \`button\` tag without error 1`] = `
<button
  class="Button secondary lg"
  type="button"
/>
`;
